SP5WWP/m17-coder/m17-coder-sym.c

```diff
@@ -42,29 +106,49 @@ int main(int argc, char* argv[])
     //printf("%d -> %d -> %d\n", 1, intrl_seq[1], intrl_seq[intrl_seq[1]]); //interleaver bijective reciprocality test, f(f(x))=x
     //return 0;
 
+    srand(time(NULL)); //seed random number generator
+    memset(key, 0, 32*sizeof(uint8_t));
+    memset(iv, 0, 16*sizeof(uint8_t));
+
     //encryption init
-    if(argc>2 && strstr(argv[1], "-K"))
+    if(argc>1 && strstr(argv[1], "-K"))
+    {
+        //hard coded key
+        for (uint8_t i = 0; i < 32; i++)
+          key[i] = 0x77;
+
         encryption=2; //AES key was passed
+    }
+
+    if(argc>1 && strstr(argv[1], "-k"))
+    {
+      // scrambler_key = atoi(argv[2]); //would prefer to get the hex input, but good enough to test with
+      scrambler_key = 0x123456;
+      encryption=1; //Scrambler key was passed
+      pn_sequence_generator();
+    }
 
     if(encryption==2)
     {
         //TODO: read key
-        AES_init_ctx(&ctx, key);
-        *((int32_t*)&iv[0])=(uint32_t)time(NULL)-(uint32_t)epoch; //timestamp
-        for(uint8_t i=4; i<4+10; i++) iv[i]=0; //10 random bytes TODO: replace with a rand() or pass through an additional arg
+        
+        // *((int32_t*)&iv[0])=(uint32_t)time(NULL)-(uint32_t)epoch; //timestamp
+        for(uint8_t i=0; i<4; i++)  iv[i] = ((uint32_t)(time(NULL)&0xFFFFFFFF)-(uint32_t)epoch) >> (24-(i*8));
+        for(uint8_t i=3; i<14; i++) iv[i] = rand() & 0xFF; //10 random bytes
     }
 
     if(fread(&(next_lsf.dst), 6, 1, stdin)<1) finished=1;
     if(fread(&(next_lsf.src), 6, 1, stdin)<1) finished=1;
     if(fread(&(next_lsf.type), 2, 1, stdin)<1) finished=1;
+    if(fread(&(next_lsf.meta), 14, 1, stdin)<1) finished=1; //this needs to be read in regardless with the current setup on m17_streamer.grc
     if(encryption==0)
     {
-        if(fread(&(next_lsf.meta), 14, 1, stdin)<1) finished=1; //read data from stdin
+        // if(fread(&(next_lsf.meta), 14, 1, stdin)<1) finished=1; //read data from stdin
     }
     else
```

Comment:
If encryption==AES then copy IV to META - that's fine. What about the seed for encryption==scrambler?

Reply:
What about the scrambler seed? Its period is reset at the start of each superframe. I don't recall anything in the specification that says its placed in the meta field or anything like that. Usually, for scramblers, you just create your keystream (enough for an entire superframe) at the start and just cycle through it over and over again until the end of transmission. Its really no different than your 'whitening' or 'randomization' sequence applied to each M17 frame, really, just an extra one that is variable based on a key value and applied to each payload portion of stream mode, resetting at the start of a new superframe.

Reply:
Oh, you are right about the seed and the META field. I must be too tired, sorry!
The LFSR should be initialized with the seed at FN=0 (I believe you do that), then used to XOR the data (also true). Shifting should occur at every frame - this is where I see the difference. You have a predefined byte array for the whole superframe, when the LFSR shuld continue 'rolling' until it basically rolls over. Otherwise there's no point in saying "seed length of 24 gives a sequence period of 16.8mil".

Reply:
I mean, if you would prefer for the scrambler to constantly generate new keystream each superframe, that can be done. Just keep in mind that the Spec is kind of vague on that point. In practice *cough cough* NXDN *cough cough*, the sequence is reset at the start of each super frame. 

To keep in mind the idea that signal can be diminished, or break up, or user can incur late entry, IF you were to use a rolling seed and constantly generate new keystream for scrambler mode, then the user would need to be able to constantly recalculate the seed value (which could be done based off of the frame number) so adding it to the META field wouldn't really be necessary.

Point B, If you were to transmit the SEED value in META, you are essentially giving away your key value, as is typically found, the LFSR initial SEED value is the key value. That's how you know how to calculate the keystream. A  "Determined Adversary" would have no issue reversing and finding a key this way.

Point C, if a known plain text (silence vector for Codec2) is ever played with scrambler at the start of transmission, then reversing a scrambler key is very trivial. Thus, while scrambler is good for obscuring a voice transmission, it is poor man's security, at best.

Reply:
The scrambler is just to prevent "the Joe next door" from decrypting the stream. No more, no less. For most cases, even phase inversion would work, but I'm not going to discuss technical skills of an average amateur radio operator.
Late joiners would have to shift the LSFR FN number of times (upon receiving that particular FN).